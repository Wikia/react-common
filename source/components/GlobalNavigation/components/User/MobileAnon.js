import React from 'react';
import PropTypes from 'prop-types';

import Button from '../../../Button';
import Avatar from '../../../Avatar';

import withRedirectUrl from '../../hocs/withRedirectUrl';

const MobileAnon = ({ data, getUrlWithRedirect }) => (
    <Button
        className="wds-global-navigation__modal-control wds-global-navigation__modal-control-anon"
        href={getUrlWithRedirect(data.signin.href, data.signin['param-name'])}
        text
    >
        <Avatar />
    </Button>
);

MobileAnon.propTypes = {
    data: PropTypes.shape().isRequired,
    getUrlWithRedirect: PropTypes.func.isRequired,
};

export default withRedirectUrl(MobileAnon);
