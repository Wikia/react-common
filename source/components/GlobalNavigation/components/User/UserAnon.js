import React from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import Dropdown from '../../../Dropdown';
import Avatar from '../../../Avatar';
import Button from '../../../Button';
import List from '../../../List';

import withRedirectUrl from '../../hocs/withRedirectUrl';

const UserToggle = ({ chevron }) => (
    <React.Fragment>
        <Avatar />
        {chevron}
    </React.Fragment>
);

UserToggle.propTypes = {
    chevron: PropTypes.node.isRequired,
};

const UserAnon = ({ data, getUrlWithRedirect }) => {
    const [t] = useTranslation();

    return (
        <Dropdown
            toggle={chevron => <UserToggle chevron={chevron} />}
            className="wds-global-navigation__user-menu wds-global-navigation__user-anon"
            contentScrollable={false}
            dropdownRightAligned
            hasShadow
            contentClassName="wds-global-navigation__dropdown-content"
        >
            <List linesBetween>
                <li>
                    <Button
                        fullwidth
                        href={getUrlWithRedirect(data.signin.href, data.signin['param-name'])}
                        rel="nofollow"
                        data-tracking-label={data.signin['tracking-label']}
                    >
                        {t(data.signin.title.key)}
                    </Button>
                </li>
                <li>
                    <div className="wds-global-navigation__user-menu-dropdown-caption">
                        Don&apos;t have an account?
                    </div>
                    <Button
                        href={getUrlWithRedirect(data.register.href, data.register['param-name'])}
                        rel="nofollow"
                        data-tracking-label={data.register['tracking-label']}
                        fullwidth
                        secondary
                    >
                        {t(data.register.title.key)}
                    </Button>
                </li>
            </List>
        </Dropdown>
    );
};

UserAnon.propTypes = {
    data: PropTypes.object.isRequired,
    getUrlWithRedirect: PropTypes.func.isRequired,
};

export default withRedirectUrl(UserAnon);
