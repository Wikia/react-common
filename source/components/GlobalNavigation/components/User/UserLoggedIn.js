import React from 'react';
import PropTypes from 'prop-types';

import Dropdown from '../../../Dropdown';
import Avatar from '../../../Avatar';
import List from '../../../List';

import LinkText from '../Link/LinkText';
import LinkLogOut from '../Link/LinkLogOut';

const UserToggle = ({ chevron, data }) => (
    <React.Fragment>
        <Avatar alt={data.username} src={data.avatar_url} />
        {chevron}
    </React.Fragment>
);

UserToggle.propTypes = {
    chevron: PropTypes.node.isRequired,
    data: PropTypes.object.isRequired,
};

/* eslint-disable react/no-array-index-key */
const UserLoggedIn = ({ data }) => (
    <Dropdown
        toggle={chevron => <UserToggle chevron={chevron} data={data} />}
        className="wds-global-navigation__user-menu wds-global-navigation__user-logged-in"
        contentScrollable={false}
        contentClassName="wds-global-navigation__dropdown-content"
        iconName="dropdown-tiny"
        hasShadow
        dropdownRightAligned
    >
        <List isLinked>
            {data.items.map((item, index) => {
                if (item.type === 'link-text') {
                    return <li key={index}><LinkText link={item} /></li>;
                }

                if (item.type === 'link-logout') {
                    return <li key={index}><LinkLogOut link={item} /></li>;
                }

                return null;
            })}
        </List>
    </Dropdown>
);

UserLoggedIn.propTypes = {
    data: PropTypes.object.isRequired,
};

export default UserLoggedIn;
