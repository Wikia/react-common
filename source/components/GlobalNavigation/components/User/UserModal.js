import React, { useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import Avatar from '../../../Avatar';

import NotificationsContext from '../../context/NotificationContext';
import I18nNamespaceContext from '../../context/I18nNamespaceContext';

import LinkLogOut from '../Link/LinkLogOut';
import NotificationsListContainer from '../Notifications/ListContainer';

const UserModal = ({ isOpen, user, track, header }) => {
    const { loadFirstPage } = useContext(NotificationsContext);
    const [t] = useTranslation(useContext(I18nNamespaceContext));
    const viewProfileLink = user.items.filter(item => item['tracking-label'] === 'account.profile')[0].href;
    const signOutModel = user.items.filter(item => item.type === 'link-logout')[0];
    const classes = classNames('wds-global-navigation__modal', 'wds-user-modal', { 'wds-is-hidden': !isOpen });

    useEffect(() => {
        if (isOpen) {
            loadFirstPage();
        }
    }, [isOpen, loadFirstPage]);

    return (
        <div className={classes}>
            <div className="wds-user-modal__info">
                <Avatar src={user.avatar_url} href={viewProfileLink} data-tracking-label="open-user-profile" />
                <p className="wds-user-modal__username">
                    {user.username}
                </p>
            </div>
            <div className="wds-user-modal__notifications-wrapper">
                <NotificationsListContainer track={track} header={t(header)} />
            </div>
            <LinkLogOut link={signOutModel} className="wds-user-modal__sign-out-container" />
        </div>
    );
};

UserModal.propTypes = {
    header: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired,
    track: PropTypes.func.isRequired,
    user: PropTypes.shape().isRequired,
};

export default UserModal;
