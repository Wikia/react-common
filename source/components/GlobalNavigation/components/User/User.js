import React from 'react';
import PropTypes from 'prop-types';

import UserLoggedIn from './UserLoggedIn';
import UserAnon from './UserAnon';

const User = ({ data }) => {
    if (data.user) {
        return <UserLoggedIn data={data.user} />;
    }

    if (data.anon) {
        return <UserAnon data={data.anon} />;
    }

    return null;
};

User.propTypes = {
    data: PropTypes.shape().isRequired,
};

export default User;
