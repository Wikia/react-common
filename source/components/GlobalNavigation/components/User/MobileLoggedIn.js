import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import Button from '../../../Button';
import Avatar from '../../../Avatar';

import NotificationsContext from '../../models/NotificationContext';

import UserModal from './UserModal';

const MobileLoggedIn = ({ user, openModal, isOpen, track }) => {
    const { unreadCount } = useContext(NotificationsContext);

    return (
        <React.Fragment>
            <Button
                className="wds-global-navigation__modal-control wds-global-navigation__modal-control-user"
                onClick={() => openModal('user')}
                text
            >
                <Avatar src={user.avatar_url} alt={user.username} />
                {
                    Boolean(unreadCount) && (
                        <span className="wds-global-navigation__avatar-notifications-counter">
                            {unreadCount > 99 ? '99+' : unreadCount}
                        </span>
                    )
                }
            </Button>
            <UserModal user={user} isOpen={isOpen} track={track} />
        </React.Fragment>
    );
};

MobileLoggedIn.propTypes = {
    isOpen: PropTypes.bool.isRequired,
    openModal: PropTypes.func.isRequired,
    track: PropTypes.func.isRequired,
    user: PropTypes.shape().isRequired,
};

export default MobileLoggedIn;
