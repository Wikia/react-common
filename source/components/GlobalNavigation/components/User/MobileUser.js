import React from 'react';
import PropTypes from 'prop-types';

import MobileAnon from './MobileAnon';
import MobileLoggedIn from './MobileLoggedIn';

const MobileUser = ({ data, openModal, isOpen, track }) => {
    if (data.user) {
        return (
            <MobileLoggedIn
                user={data.user}
                openModal={openModal}
                isOpen={isOpen}
                track={track}
            />
        );
    }

    if (data.anon) {
        return <MobileAnon data={data.anon} />;
    }

    return null;
};

MobileUser.propTypes = {
    data: PropTypes.shape().isRequired,
    isOpen: PropTypes.bool.isRequired,
    openModal: PropTypes.func.isRequired,
    track: PropTypes.func.isRequired,
};

export default MobileUser;
