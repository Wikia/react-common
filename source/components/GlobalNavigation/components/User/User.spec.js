import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';

import User from './User';

const defaultProps = {
    data: {},
};

function renderComponent(props) {
    const computedProps = merge({}, defaultProps, props);

    return shallow(<User {...computedProps} />);
}

test('User renders correctly for user', () => {
    expect(renderComponent({ data: { user: { mockedUser: 'mocked-user-value' } } })).toMatchSnapshot();
});

test('User renders correctly for anon', () => {
    expect(renderComponent({ data: { anon: { mockedAnon: 'mocked-anon-value' } } })).toMatchSnapshot();
});

test('User returns null when no data is provided', () => {
    expect(renderComponent()).toMatchSnapshot();
});
