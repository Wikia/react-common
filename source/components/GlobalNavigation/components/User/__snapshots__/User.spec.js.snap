// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User renders correctly for anon 1`] = `
<ComponentWrapper
  data={
    Object {
      "mockedAnon": "mocked-anon-value",
    }
  }
/>
`;

exports[`User renders correctly for user 1`] = `
<UserLoggedIn
  data={
    Object {
      "mockedUser": "mocked-user-value",
    }
  }
/>
`;

exports[`User returns null when no data is provided 1`] = `""`;
